doc/development/powershell.md

```diff
@@ -0,0 +1,165 @@
+This page contains guidelines for developing or updating powershell scripts used by the Azure SDK engineering system.
+
+Table of Contents
+=================
+
+* [TLDR](#tldr)
+* [Structure](#structure)
+* [Functionality](#functionality)
+* [Style](#style)
+* [Testing](#testing)
+   * [Unit Testing](#unit-testing)
+      * [Running Pester Tests](#running-pester-tests)
+   * [Local/Pipeline Functional Testing](#localpipeline-functional-testing)
+
+## TLDR
+
+_(i.e. too long, didn't read)_
+
+**DO**
+
+- Try to write scripts that can be run locally
+- Write unit tests for your scripts
+- Handle exit codes from external programs
+- Use `Set-StrictMode -Version 4` where possible
+- Define your data structures as much as possible via composition, or unit test or comment examples.
+- Use an editor extension or `Invoke-ScriptAnalyzer` to analyze code.
+
+**DON'T**
+
+- Write medium-large scripts inline to azure pipelines yaml (unless absolutely necessary)
+- Write code without putting logic into functions. Scripts grow in size and complexity over time, and if they start out un-organized, they tend to stay un-organized.
+
+## Structure
+
+When structuring a powershell cmdlet or collection of scripts, it is important to think about how easy that code will be to re-use and to test (primarily unit tests and/or running locally). 
+
+- When writing a powershell cmdlet file, keep the file as small as possible, or organize your code into functions. Scripts tend to grow in size and complexity over time with incremental edits from many authors, and it is much easier to understand, test and refactor code when it is written in a modular way from the start.
+- Isolate calls to dependencies inside their own cmdlet files or within functions. This makes it easier to abstract these calls for testing.
+- Avoid global state where possible. Powershell has very liberal scoping rules, and it is easy to take dependencies on variables that are defined outside a code block, which can lead to hard to debug issues. Where possible, only treat script parameters and variables defined at the top of the file as read-only global variables.
+- Try not to rely on exporting your code as a powershell module. Favor [dot sourcing](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_scripts?view=powershell-7.2#script-scope-and-dot-sourcing) when possible. There are lots of downsides to working with module imports in local development settings that make them not worth it. This guideline some limitations, e.g. if you want to test class instance assertions via Pester, you cannot export the class types without using module imports.
```

Comment:
Modules also have benefits like disambiguation and I've used them in countless automations (just use an explicit path if `$env:PSModulePath` is a problem), but 🤷‍♂️.

But there's a grammatical mistake here toward the end: "This guideline some limitations". I think you're trying to say "This guideline lists some limitations"

...though I don't see what limitations you list. Scoping all dot-sourced variables to the global scope isn't often a good thing. People can end up overwriting variables they didn't expect to, where as with modules they are scoped to the module (or function, if in a script block) by default unless explicitly added to the `global:` scope. You have options then.

Reply:
Our main pain with modules is you can't easily reload them on changes as it will just skip the import if you've already loaded it. You need to restart the shell session.

Reply:
I'll rephrase the last part you mention. What I was trying to say was "by following this no module guideline you will lose some nice things."

Reply:
I should clarify re: global, I meant more like setting a module constant as opposed to `$global` scope. I forgot that was a more loaded term in powershell.

```
$MY_MODULE_CONSTANT = "foo"
``

Reply:
> Our main pain with modules is you can't easily reload them on changes as it will just skip the import if you've already loaded it. You need to restart the shell session.

You just use `Remove-Module <name>`. I've written countless modules and this has been the primary way of testing changes. In fact, interactively, you can use `#remo<tab>` to execute a previous command (like bash's `^`). So a session ends up looking like:

```pwsh
vim mod.psm1
ipmo ./mod.psm1
# oops
remove-module mod
vim mod.psm1
#ip<tab>
# repeat
#rem<tab>
# ...
```

That said, I'm not married to the use of modules here. Just that there seems to be a lot of misunderstanding - like that they can't be unloaded. On the contrary, with everything (variables, functions, etc) scoped to the module by default, removing the module cleans all that up. Not true with dot sourcing a script which is a one-way deal. Re-sourcing it will overwrite those variables, but there's no way to simply clean up like with a module. Modules also allow you to disambiguate like `mod\select-object` to avoid confusion with the built-in `select-object`.

Reply:
That's a fair point, and if it addresses our main issue, then I would prefer to start using modules so we can more strictly define our data structures. @weshaggard @scbedd were there any other primary module concerns you can think of?

Reply:
Module identifiers can be unloaded but types cannot as they are loaded into the global address space of PS and often times require a restart of the shell. That is also why I often avoid classes in scripts as well. I don't know how many times I have banged my head against the wall trying to figure out why my changes to a class or module are not being picked up correctly. From my experience modules cause more trouble than they are worth and for limited to no benefits that I've seen in practice.

Reply:
Using classes and dot-sourcing files would be no different, except that *everything* is in the global scope when dot-sourcing by design.